---
 drivers/net/mlx4/en_params.c |    2 --
 1 file changed, 2 deletions(-)

Index: ofed_kernel/drivers/net/mlx4/en_params.c
===================================================================
--- ofed_kernel.orig/drivers/net/mlx4/en_params.c
+++ ofed_kernel/drivers/net/mlx4/en_params.c
@@ -593,8 +593,6 @@ const struct ethtool_ops mlx4_en_ethtool
 	.set_pauseparam = mlx4_en_set_pauseparam,
 	.get_ringparam = mlx4_en_get_ringparam,
 	.set_ringparam = mlx4_en_set_ringparam,
-	.get_flags = ethtool_op_get_flags,
-	.set_flags = ethtool_op_set_flags,
 };
 
 
