---
 drivers/net/cxgb3/version.h |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: ofed_kernel/drivers/net/cxgb3/version.h
===================================================================
--- ofed_kernel.orig/drivers/net/cxgb3/version.h
+++ ofed_kernel/drivers/net/cxgb3/version.h
@@ -35,7 +35,7 @@
 #define DRV_DESC "Chelsio T3 Network Driver"
 #define DRV_NAME "cxgb3"
 /* Driver version */
-#define DRV_VERSION "1.0-ko"
+#define DRV_VERSION "1.0-ofed"
 
 /* Firmware version */
 #define FW_VERSION_MAJOR 7
