diff -up ofed_kernel-2.6.18-EL5.1/drivers/infiniband/hw/ipath/ipath_verbs.c c/drivers/infiniband/hw/ipath/ipath_verbs.c
--- ofed_kernel-2.6.18-EL5.1/drivers/infiniband/hw/ipath/ipath_verbs.c	2009-03-03 17:55:38.000000000 -0800
+++ c/drivers/infiniband/hw/ipath/ipath_verbs.c	2009-03-05 16:36:44.000000000 -0800
@@ -1218,6 +1218,7 @@ static int ipath_verbs_send_dma(struct i
 	tx->txreq.map_addr = piobuf;
 	tx->txreq.flags |= IPATH_SDMA_TXREQ_F_FREEBUF;
 	tx->txreq.sg_count = 1;
+	memcpy(&tx->hdr.hdr, hdr, hdrwords << 2);
 
 	*piobuf++ = (__force u32) cpu_to_le32(plen);
 	*piobuf++ = (__force u32) cpu_to_le32(control);
