diff --git a/drivers/infiniband/hw/ipath/ipath_user_sdma.c b/drivers/infiniband/hw/ipath/ipath_user_sdma.c
index 82d9a0b..a4cc9b9 100644
--- a/drivers/infiniband/hw/ipath/ipath_user_sdma.c
+++ b/drivers/infiniband/hw/ipath/ipath_user_sdma.c
@@ -712,6 +712,8 @@ static int ipath_user_sdma_push_pkts(struct ipath_devdata *dd,
 	int ret = 0;
 	unsigned long flags;
 	u16 tail;
+	u8 generation;
+	u64 descq_added;
 
 	if (list_empty(pktlist))
 		return 0;
@@ -721,6 +723,10 @@ static int ipath_user_sdma_push_pkts(struct ipath_devdata *dd,
 
 	spin_lock_irqsave(&dd->ipath_sdma_lock, flags);
 
+	/* keep a copy for restoring purposes in case of problems */
+	generation = dd->ipath_sdma_generation;
+	descq_added = dd->ipath_sdma_descq_added;
+
 	if (unlikely(dd->ipath_sdma_status & IPATH_SDMA_ABORT_MASK)) {
 		ret = -ECOMM;
 		goto unlock;
@@ -784,6 +790,10 @@ unlock_check_tail:
 	}
 
 unlock:
+	if (unlikely(ret < 0)) {
+		dd->ipath_sdma_generation = generation;
+		dd->ipath_sdma_descq_added = descq_added;
+	}
 	spin_unlock_irqrestore(&dd->ipath_sdma_lock, flags);
 
 	return ret;
