diff --git old/drivers/infiniband/hw/ipath/ipath_iba6120.c new/drivers/infiniband/hw/ipath/ipath_iba6120.c
index ba1f51d..ac38ca8 100644
--- old/drivers/infiniband/hw/ipath/ipath_iba6120.c
+++ new/drivers/infiniband/hw/ipath/ipath_iba6120.c
@@ -709,6 +709,10 @@ static void ipath_pe_init_hwerrors(struct ipath_devdata *dd)
 		 */
 		val &= ~INFINIPATH_HWE_PCIEBUSPARITYRADM;
 	}
+
+	/* avoid some intel cpu's speculative read freeze mode issue */
+	val &= ~(INFINIPATH_HWE_TXEMEMPARITYERR_PIOBUF
+	   << INFINIPATH_HWE_TXEMEMPARITYERR_SHIFT);
 	dd->ipath_hwerrmask = val;
 }
