--- a/drivers/infiniband/hw/ipath/ipath_ruc.c	2009-01-22 11:45:13.000000000 -0800
+++ b/drivers/infiniband/hw/ipath/ipath_ruc.c	2009-01-23 11:54:46.000000000 -0800
@@ -660,12 +660,14 @@ void ipath_do_send(unsigned long data)
 {
 	struct ipath_qp *qp = (struct ipath_qp *)data;
 	struct ipath_ibdev *dev = to_idev(qp->ibqp.device);
+	struct ipath_devdata *dd = dev->dd;
 	int (*make_req)(struct ipath_qp *qp);
 	unsigned long flags;
 
 	if ((qp->ibqp.qp_type == IB_QPT_RC ||
 	     qp->ibqp.qp_type == IB_QPT_UC) &&
-	    qp->remote_ah_attr.dlid == dev->dd->ipath_lid) {
+	    (qp->remote_ah_attr.dlid & ~((1 << dd->ipath_lmc) - 1)) ==
+	    dd->ipath_lid) {
 		ipath_ruc_loopback(qp);
 		goto bail;
 	}
