diff -up ofed_kernel-2.6.18-EL5.1/drivers/infiniband/hw/ipath/ipath_user_sdma.c c/drivers/infiniband/hw/ipath/ipath_user_sdma.c
--- ofed_kernel-2.6.18-EL5.1/drivers/infiniband/hw/ipath/ipath_user_sdma.c	2009-03-03 17:55:38.000000000 -0800
+++ c/drivers/infiniband/hw/ipath/ipath_user_sdma.c	2009-03-05 15:53:17.000000000 -0800
@@ -672,8 +672,8 @@ static inline __le64 ipath_sdma_make_fir
 
 static inline __le64 ipath_sdma_make_last_desc0(__le64 descq)
 {
-					      /* last */  /* dma head */
-	return descq | __constant_cpu_to_le64(1ULL << 11 | 1ULL << 13);
+					      /* last */
+	return descq | __constant_cpu_to_le64(1ULL << 11);
 }
 
 static inline __le64 ipath_sdma_make_desc1(u64 addr)
