diff --git a/drivers/infiniband/hw/ipath/ipath_driver.c b/drivers/infiniband/hw/ipath/ipath_driver.c
index ad0aab6..7cdf7f1 100644
--- a/drivers/infiniband/hw/ipath/ipath_driver.c
+++ b/drivers/infiniband/hw/ipath/ipath_driver.c
@@ -1102,7 +1102,7 @@ struct sk_buff *ipath_alloc_skb(struct ipath_devdata *dd,
 	 * key header.  In order to keep everything dword aligned,
 	 * we'll reserve 4 bytes.
 	 */
-	len = dd->ipath_ibmaxlen + 4;
+	len = dd->ipath_init_ibmaxlen + 4;
 
 	if (dd->ipath_flags & IPATH_4BYTE_TID) {
 		/* We need a 2KB multiple alignment, and there is no way
@@ -2506,7 +2506,7 @@ void ipath_free_pddata(struct ipath_devdata *dd, struct ipath_portdata *pd)
 		for (e = 0; e < dd->ipath_p0_rcvegrcnt; e++)
 			if (skbinfo[e].skb) {
 				pci_unmap_single(dd->pcidev, skbinfo[e].phys,
-						 dd->ipath_ibmaxlen,
+						 dd->ipath_init_ibmaxlen,
 						 PCI_DMA_FROMDEVICE);
 				dev_kfree_skb(skbinfo[e].skb);
 			}
diff --git a/drivers/infiniband/hw/ipath/ipath_init_chip.c b/drivers/infiniband/hw/ipath/ipath_init_chip.c
index 3e5baa4..39afba1 100644
--- a/drivers/infiniband/hw/ipath/ipath_init_chip.c
+++ b/drivers/infiniband/hw/ipath/ipath_init_chip.c
@@ -126,7 +126,7 @@ static int create_port0_egr(struct ipath_devdata *dd)
 		dd->ipath_port0_skbinfo[e].phys =
 		  ipath_map_single(dd->pcidev,
 				   dd->ipath_port0_skbinfo[e].skb->data,
-				   dd->ipath_ibmaxlen, PCI_DMA_FROMDEVICE);
+				   dd->ipath_init_ibmaxlen, PCI_DMA_FROMDEVICE);
 		dd->ipath_f_put_tid(dd, e + (u64 __iomem *)
 				    ((char __iomem *) dd->ipath_kregbase +
 				     dd->ipath_rcvegrbase),
