--- a/drivers/infiniband/hw/ipath/ipath_rc.c	2009-03-03 13:55:04.000000000 -0800
+++ b/drivers/infiniband/hw/ipath/ipath_rc.c	2009-03-03 13:38:51.000000000 -0800
@@ -940,6 +940,7 @@ void ipath_rc_send_complete(struct ipath
 			wc.wr_id = wqe->wr.wr_id;
 			wc.status = IB_WC_SUCCESS;
 			wc.opcode = ib_ipath_wc_opcode[wqe->wr.opcode];
+			wc.byte_len = wqe->length;
 			wc.qp = &qp->ibqp;
 			ipath_cq_enter(to_icq(qp->ibqp.send_cq), &wc, 0);
 		}
@@ -1086,6 +1087,7 @@ static int do_rc_ack(struct ipath_qp *qp
 				wc.wr_id = wqe->wr.wr_id;
 				wc.status = IB_WC_SUCCESS;
 				wc.opcode = ib_ipath_wc_opcode[wqe->wr.opcode];
+				wc.byte_len = wqe->length;
 				wc.qp = &qp->ibqp;
 				ipath_cq_enter(to_icq(qp->ibqp.send_cq), &wc,
 						0);
