--- a/fs/nfsd/nfsctl.c	2009-03-29 18:53:02.000000000 -0500
+++ b/fs/nfsd/nfsctl.c	2009-03-31 10:44:32.000000000 -0500
@@ -657,6 +657,9 @@ static ssize_t __write_ports(struct file
 					/* Give a reasonable perror msg for
 					 * bad transport string */
 					err = -EPROTONOSUPPORT;
+				if (err >= 0)
+					lockd_up(0);
+				nfsd_serv->sv_nrthreads--;
 			}
 			return err < 0 ? err : 0;
 		}
@@ -679,6 +682,7 @@ static ssize_t __write_ports(struct file
 					svc_close_xprt(xprt);
 					svc_xprt_put(xprt);
 					err = 0;
+					lockd_down();
 				} else
 					err = -ENOTCONN;
 			}
