---
 drivers/infiniband/hw/cxgb3/iwch_cm.c |    2 ++
 1 file changed, 2 insertions(+)

Index: ofed_kernel/drivers/infiniband/hw/cxgb3/iwch_cm.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/cxgb3/iwch_cm.c
+++ ofed_kernel/drivers/infiniband/hw/cxgb3/iwch_cm.c
@@ -47,6 +47,7 @@
 #include "iwch_provider.h"
 #include "iwch_cm.h"
 
+#ifdef DEBUG
 static char *states[] = {
 	"idle",
 	"listen",
@@ -62,6 +63,7 @@ static char *states[] = {
 	"dead",
 	NULL,
 };
+#endif
 
 int peer2peer = 0;
 module_param(peer2peer, int, 0644);
