diff --git a/drivers/infiniband/hw/nes/nes_nic.c b/drivers/infiniband/hw/nes/nes_nic.c
index 12d9cc6..95f5a3f 100644
--- a/drivers/infiniband/hw/nes/nes_nic.c
+++ b/drivers/infiniband/hw/nes/nes_nic.c
@@ -1549,6 +1549,11 @@ static void nes_netdev_vlan_rx_register(struct net_device *netdev, struct vlan_g
 	spin_unlock_irqrestore(&nesadapter->phy_lock, flags);
 }
 
+static void nes_netdev_vlan_rx_kill_vid(struct net_device *netdev, 
+					unsigned short vid)
+{
+}
+
 
 /**
  * nes_netdev_init - initialize network device
@@ -1598,6 +1603,7 @@ struct net_device *nes_netdev_init(struct nes_device *nesdev,
 	nes_debug(NES_DBG_INIT, "Enabling VLAN Insert/Delete.\n");
 	netdev->features |= NETIF_F_HW_VLAN_TX | NETIF_F_HW_VLAN_RX;
 	netdev->vlan_rx_register = nes_netdev_vlan_rx_register;
+	netdev->vlan_rx_kill_vid = nes_netdev_vlan_rx_kill_vid;
 
 	/* Fill in the port structure */
 	nesvnic->netdev = netdev;
-- 
1.5.3.3

