From a5b4e2f59bc0eec5d53f93afff72fc2bb67cc764 Mon Sep 17 00:00:00 2001
From: Vladimir Sokolovsky <vlad@mellanox.co.il>
Date: Sun, 27 Jan 2008 15:58:24 +0200
Subject: [PATCH] backport SRPT: add include <linux/scatterlist.h> for RHEL4.0U[4-6].

Signed-off-by: Vladimir Sokolovsky <vlad@mellanox.co.il>
---
 drivers/infiniband/ulp/srpt/ib_srpt.c |    1 +
 1 file changed, 1 insertion(+)

Index: ofed_kernel/drivers/infiniband/ulp/srpt/ib_srpt.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/ulp/srpt/ib_srpt.c
+++ ofed_kernel/drivers/infiniband/ulp/srpt/ib_srpt.c
@@ -37,6 +37,7 @@
 #include <linux/err.h>
 #include <linux/string.h>
 #include <linux/kthread.h>
+#include <linux/scatterlist.h>
 
 #include <asm/atomic.h>
 
