---
 drivers/infiniband/ulp/sdp/sdp_main.c |   21 ---------------------
 1 file changed, 21 deletions(-)

Index: ofed_kernel/drivers/infiniband/ulp/sdp/sdp_main.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/ulp/sdp/sdp_main.c
+++ ofed_kernel/drivers/infiniband/ulp/sdp/sdp_main.c
@@ -39,27 +39,6 @@
  *  2 of the License, or(at your option) any later version.
  */
 
-#if defined(__ia64__)
-/* csum_partial_copy_from_user is not exported on ia64.
-   We don't really need it for SDP - skb_copy_to_page happens to call it
-   but for SDP HW checksum is always set, so ... */
-
-#include <linux/errno.h>
-#include <linux/types.h>
-#include <asm/checksum.h>
-
-static inline
-unsigned int csum_partial_copy_from_user_new (const char *src, char *dst,
-						 int len, unsigned int sum,
-						 int *errp)
-{
-	*errp = -EINVAL;
-	return 0;
-}
-
-#define csum_partial_copy_from_user csum_partial_copy_from_user_new
-#endif
-
 #include <linux/tcp.h>
 #include <asm/ioctls.h>
 #include <linux/workqueue.h>
