diff --git a/drivers/net/cxgb3/adapter.h b/drivers/net/cxgb3/adapter.h
index 21dad82..1c4e828 100644
--- a/drivers/net/cxgb3/adapter.h
+++ b/drivers/net/cxgb3/adapter.h
@@ -207,6 +207,7 @@ struct adapter {
 	struct list_head adapter_list;
 	void __iomem *regs;
 	struct pci_dev *pdev;
+	u32 saved_pci_state[16];
 	unsigned long registered_device_map;
 	unsigned long open_device_map;
 	unsigned long flags;
diff --git a/drivers/net/cxgb3/t3_hw.c b/drivers/net/cxgb3/t3_hw.c
index 533fc74..f6c000e 100644
--- a/drivers/net/cxgb3/t3_hw.c
+++ b/drivers/net/cxgb3/t3_hw.c
@@ -3534,7 +3534,7 @@ static int t3_reset_adapter(struct adapter *adapter)
 	uint16_t devid = 0;
 
 	if (save_and_restore_pcie)
-		pci_save_state(adapter->pdev);
+ 		pci_save_state(adapter->pdev, adapter->saved_pci_state);
 	t3_write_reg(adapter, A_PL_RST, F_CRSTWRM | F_CRSTWRMMODE);
 
 	/*
@@ -3552,7 +3552,7 @@ static int t3_reset_adapter(struct adapter *adapter)
 		return -1;
 
 	if (save_and_restore_pcie)
-		pci_restore_state(adapter->pdev);
+ 		pci_restore_state(adapter->pdev, adapter->saved_pci_state);
 	return 0;
 }
 
@@ -3688,8 +3688,6 @@ int t3_prep_adapter(struct adapter *adapter, const struct adapter_info *ai,
 
 		memcpy(adapter->port[i]->dev_addr, hw_addr,
 		       ETH_ALEN);
-		memcpy(adapter->port[i]->perm_addr, hw_addr,
-		       ETH_ALEN);
 		init_link_config(&p->link_config, p->phy.caps);
 		p->phy.ops->power_down(&p->phy, 1);
 		if (!(p->phy.caps & SUPPORTED_IRQ))
