Index: ofed_kernel/drivers/infiniband/ulp/ipoib/ipoib_ethtool.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/ulp/ipoib/ipoib_ethtool.c
+++ ofed_kernel/drivers/infiniband/ulp/ipoib/ipoib_ethtool.c
@@ -100,14 +100,9 @@ static void ipoib_get_strings(struct net
 	}
 }
 
-static int ipoib_get_sset_count(struct net_device *dev, int sset)
+static int ipoib_get_stats_count(struct net_device *dev)
 {
-	switch (sset) {
-	case ETH_SS_STATS:
-		return ARRAY_SIZE(ipoib_stats_keys);
-	default:
-		return -EOPNOTSUPP;
-	}
+	return ARRAY_SIZE(ipoib_stats_keys);
 }
 
 static void ipoib_get_ethtool_stats(struct net_device *dev,
@@ -132,10 +127,8 @@ static const struct ethtool_ops ipoib_et
 	.get_tso		= ethtool_op_get_tso,
 	.get_coalesce		= ipoib_get_coalesce,
 	.set_coalesce		= ipoib_set_coalesce,
-	.get_flags		= ethtool_op_get_flags,
-	.set_flags		= ethtool_op_set_flags,
 	.get_strings		= ipoib_get_strings,
-	.get_sset_count		= ipoib_get_sset_count,
+	.get_stats_count 	= ipoib_get_stats_count,
 	.get_ethtool_stats	= ipoib_get_ethtool_stats,
 };
 
