---
 drivers/infiniband/core/addr.c |    6 ++++++
 1 file changed, 6 insertions(+)

Index: ofed_kernel-2.6.9_U7/drivers/infiniband/core/addr.c
===================================================================
--- ofed_kernel-2.6.9_U7.orig/drivers/infiniband/core/addr.c	2009-06-10 10:14:51.000000000 +0300
+++ ofed_kernel-2.6.9_U7/drivers/infiniband/core/addr.c	2009-06-10 10:39:59.000000000 +0300
@@ -32,6 +32,12 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
+#include <linux/if.h>
+#include <linux/netdevice.h>
+#include <linux/timer.h>
+
+/* XXX this define is the same for all architectures in RHAS4.0. */
+#define ECANCELED 125
 
 #include <linux/mutex.h>
 #include <linux/inetdevice.h>
@@ -40,6 +46,7 @@
 #include <net/arp.h>
 #include <net/neighbour.h>
 #include <net/route.h>
+#include <net/ipv6.h>
 #include <net/addrconf.h>
 #include <net/ip6_route.h>
 #include <rdma/ib_addr.h>
