---
 drivers/infiniband/core/sysfs.c |    4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

Index: ofed_kernel/drivers/infiniband/core/sysfs.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/sysfs.c
+++ ofed_kernel/drivers/infiniband/core/sysfs.c
@@ -511,7 +511,6 @@ static int add_port(struct ib_device *de
 {
 	struct ib_port *p;
 	struct ib_port_attr attr;
-	struct attribute *kobj_attr;
 	int i;
 	int ret;
 
@@ -557,8 +556,7 @@ static int add_port(struct ib_device *de
 
 	list_add_tail(&p->kobj.entry, &device->port_list);
 
-	kobj_attr->name = p->kobj.name;
-	kobject_uevent(&p->kobj, KOBJ_ADD, kobj_attr);
+	kobject_hotplug("add", &p->kobj);
 	return 0;
 
 err_free_pkey:
