commit 7de58f29ecaf92a7d54d538e5ed91e3cc12e8186
Author: root <root@almus.(none)>
Date:   Tue Jul 7 07:07:30 2009 -0700

    cxgb3_0200_fl1_mpage_fix.patch

diff --git a/drivers/net/cxgb3/sge.c b/drivers/net/cxgb3/sge.c
index 8859f61..1a9c065 100644
--- a/drivers/net/cxgb3/sge.c
+++ b/drivers/net/cxgb3/sge.c
@@ -59,8 +59,8 @@
 #define FL0_PG_CHUNK_SIZE  2048
 #define FL0_PG_ORDER 0
 #define FL0_PG_ALLOC_SIZE (PAGE_SIZE << FL0_PG_ORDER)
-#define FL1_PG_CHUNK_SIZE (PAGE_SIZE > 8192 ? 16384 : 8192)
-#define FL1_PG_ORDER (PAGE_SIZE > 8192 ? 0 : 1)
+#define FL1_PG_CHUNK_SIZE  0
+#define FL1_PG_ORDER  0
 #define FL1_PG_ALLOC_SIZE (PAGE_SIZE << FL1_PG_ORDER)
 
 #define SGE_RX_DROP_THRES 16
