---
 fs/nfs/nfsroot.c |    2 ++
 1 file changed, 2 insertions(+)

Index: ofa_kernel-1.4/fs/nfs/nfsroot.c
===================================================================
--- ofa_kernel-1.4.orig/fs/nfs/nfsroot.c
+++ ofa_kernel-1.4/fs/nfs/nfsroot.c
@@ -376,6 +376,7 @@ static int __init root_nfs_init(void)
 }
 
 
+#ifndef MODULE
 /*
  *  Parse NFS server and directory information passed on the kernel
  *  command line.
@@ -396,6 +397,7 @@ static int __init nfs_root_setup(char *l
 }
 
 __setup("nfsroot=", nfs_root_setup);
+#endif
 
 /***************************************************************************
 
