Index: linux-2.6-work/drivers/infiniband/core/verbs.c
===================================================================
--- linux-2.6-work.orig/drivers/infiniband/core/verbs.c
+++ linux-2.6-work/drivers/infiniband/core/verbs.c
@@ -371,8 +371,8 @@ EXPORT_SYMBOL(ib_create_qp);
 
 static const struct {
 	int			valid;
-	enum ib_qp_attr_mask	req_param[IB_QPT_RAW_ETY + 1];
-	enum ib_qp_attr_mask	opt_param[IB_QPT_RAW_ETY + 1];
+	enum ib_qp_attr_mask	req_param[IB_QPT_RAW_ETH + 1];
+	enum ib_qp_attr_mask	opt_param[IB_QPT_RAW_ETH + 1];
 } qp_state_table[IB_QPS_ERR + 1][IB_QPS_ERR + 1] = {
 	[IB_QPS_RESET] = {
 		[IB_QPS_RESET] = { .valid = 1 },
@@ -382,6 +382,7 @@ static const struct {
 				[IB_QPT_UD]  = (IB_QP_PKEY_INDEX		|
 						IB_QP_PORT			|
 						IB_QP_QKEY),
+				[IB_QPT_RAW_ETH] = IB_QP_PORT,
 				[IB_QPT_UC]  = (IB_QP_PKEY_INDEX		|
 						IB_QP_PORT			|
 						IB_QP_ACCESS_FLAGS),
@@ -1004,11 +1005,15 @@ int ib_attach_mcast(struct ib_qp *qp, un
 
 	switch (rdma_node_get_transport(qp->device->node_type)) {
 	case RDMA_TRANSPORT_IB:
-		if (gid->raw[0] != 0xff || qp->qp_type != IB_QPT_UD)
+		if (qp->qp_type == IB_QPT_RAW_ETH) {
+			/* In raw Etherent mgids the 63 msb's should be 0 */
+			if (gid->global.subnet_prefix & cpu_to_be64(~1ULL))
+				return -EINVAL;
+		} else if (gid->raw[0] != 0xff || qp->qp_type != IB_QPT_UD)
 			return -EINVAL;
 		break;
 	case RDMA_TRANSPORT_IWARP:
-		if (qp->qp_type != IB_QPT_RAW_ETY)
+		if (qp->qp_type != IB_QPT_RAW_ETH)
 			return -EINVAL;
 		break;
 	}
@@ -1023,11 +1028,15 @@ int ib_detach_mcast(struct ib_qp *qp, un
 
 	switch (rdma_node_get_transport(qp->device->node_type)) {
 	case RDMA_TRANSPORT_IB:
-		if (gid->raw[0] != 0xff || qp->qp_type != IB_QPT_UD)
+		if (qp->qp_type == IB_QPT_RAW_ETH) {
+			/* In raw Etherent mgids the 63 msb's should be 0 */
+			if (gid->global.subnet_prefix & cpu_to_be64(~1ULL))
+				return -EINVAL;
+		} else if (gid->raw[0] != 0xff || qp->qp_type != IB_QPT_UD)
 			return -EINVAL;
 		break;
 	case RDMA_TRANSPORT_IWARP:
-		if (qp->qp_type != IB_QPT_RAW_ETY)
+		if (qp->qp_type != IB_QPT_RAW_ETH)
 			return -EINVAL;
 		break;
 	}
Index: linux-2.6-work/include/rdma/ib_verbs.h
===================================================================
--- linux-2.6-work.orig/include/rdma/ib_verbs.h
+++ linux-2.6-work/include/rdma/ib_verbs.h
@@ -571,7 +571,8 @@ enum ib_qp_type {
 	IB_QPT_UD,
 	IB_QPT_XRC,
 	IB_QPT_RAW_IPV6,
-	IB_QPT_RAW_ETY
+	IB_QPT_RAW_ETY,
+	IB_QPT_RAW_ETH
 };
 
 enum ib_qp_create_flags {
