Index: ofed_kernel-fixes/drivers/infiniband/ulp/ipoib/ipoib_main.c
===================================================================
--- ofed_kernel-fixes.orig/drivers/infiniband/ulp/ipoib/ipoib_main.c	2010-08-30 14:39:52.000000000 +0300
+++ ofed_kernel-fixes/drivers/infiniband/ulp/ipoib/ipoib_main.c	2010-08-30 14:55:48.483462007 +0300
@@ -1265,6 +1265,7 @@ static struct net_device *ipoib_add_port
 		goto alloc_mem_failed;
 
 	SET_NETDEV_DEV(priv->dev, hca->dma_device);
+	priv->dev->dev_id = port - 1;
 
 	if (!ib_query_port(hca, port, &attr))
 		priv->max_ib_mtu = ib_mtu_enum_to_int(attr.max_mtu);
