From 933fd51d55ab5ab9f21ad140ca2e9d337d6b117b Mon Sep 17 00:00:00 2001
From: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
Date: Thu, 17 Jul 2008 16:51:18 +0300
Subject: [PATCH] mlx4_ib: Don't register ib device if no ib ports are supported

Signed-off-by: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
---
 drivers/infiniband/hw/mlx4/qp.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: ofed_kernel/drivers/infiniband/hw/mlx4/qp.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/mlx4/qp.c
+++ ofed_kernel/drivers/infiniband/hw/mlx4/qp.c
@@ -1068,7 +1068,7 @@ static int __mlx4_ib_modify_qp(struct ib
 
 	if (attr_mask & IB_QP_ALT_PATH) {
 		if (attr->alt_port_num == 0 ||
-		    attr->alt_port_num > dev->dev->caps.num_ports) {
+		    attr->alt_port_num > dev->num_ports) {
 			mlx4_ib_dbg("qpn 0x%x: invalid alternate port num (%d)",
 				    ibqp->qp_num, attr->alt_port_num);
 			goto out;
@@ -1276,7 +1276,7 @@ int mlx4_ib_modify_qp(struct ib_qp *ibqp
 	}
 
 	if ((attr_mask & IB_QP_PORT) &&
-	    (attr->port_num == 0 || attr->port_num > dev->dev->caps.num_ports)) {
+	    (attr->port_num == 0 || attr->port_num > dev->num_ports)) {
 		mlx4_ib_dbg("qpn 0x%x: invalid port number (%d) specified "
 			    "for transition %d to %d. qp_type %d",
 			    ibqp->qp_num, attr->port_num, cur_state,
