From 0eef6f53283ef16811626d7b93c26de7877ff2e1 Mon Sep 17 00:00:00 2001
From: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
Date: Mon, 26 Jul 2010 17:40:43 +0300
Subject: [PATCH]  mlx4_en: validate vlan dev existance before accessing it when change tso

Signed-off-by: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
---
 drivers/net/mlx4/en_ethtool.c |   12 ++++++++----
 1 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/net/mlx4/en_ethtool.c b/drivers/net/mlx4/en_ethtool.c
index 7035729..5fb5d97 100644
--- a/drivers/net/mlx4/en_ethtool.c
+++ b/drivers/net/mlx4/en_ethtool.c
@@ -94,8 +94,10 @@ static int mlx4_en_set_tso(struct net_device *dev, u32 data)
 			struct net_device *vdev;
 			for (i = 0; i < VLAN_GROUP_ARRAY_LEN; i++) {
 				vdev = vlan_group_get_device(priv->vlgrp, i);
-				vdev->features |= (NETIF_F_TSO | NETIF_F_TSO6);
-				vlan_group_set_device(priv->vlgrp, i, vdev);
+				if (vdev) {
+					vdev->features |= (NETIF_F_TSO | NETIF_F_TSO6);
+					vlan_group_set_device(priv->vlgrp, i, vdev);
+				}
 			}
 		}
 #endif
@@ -109,8 +111,10 @@ static int mlx4_en_set_tso(struct net_device *dev, u32 data)
 			struct net_device *vdev;
 			for (i = 0; i < VLAN_GROUP_ARRAY_LEN; i++) {
 				vdev = vlan_group_get_device(priv->vlgrp, i);
-				vdev->features &= ~(NETIF_F_TSO | NETIF_F_TSO6);
-				vlan_group_set_device(priv->vlgrp, i, vdev);
+				if (vdev) {
+					vdev->features &= ~(NETIF_F_TSO | NETIF_F_TSO6);
+					vlan_group_set_device(priv->vlgrp, i, vdev);
+				}
 			}
 		}
 #endif
-- 
1.7.1.1

