From 1acbcbe34ac561490f369a521b2f48449aea231d Mon Sep 17 00:00:00 2001
From: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
Date: Sun, 14 Feb 2010 09:44:50 +0200
Subject: [PATCH] mlx4_en: Destoy allocator only when using frags

Signed-off-by: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
---
 drivers/net/mlx4/en_rx.c |    7 ++++---
 1 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/net/mlx4/en_rx.c b/drivers/net/mlx4/en_rx.c
index 5c71c4f..716ebc8 100644
--- a/drivers/net/mlx4/en_rx.c
+++ b/drivers/net/mlx4/en_rx.c
@@ -290,7 +290,7 @@ static int mlx4_en_fill_rx_buffers(struct mlx4_en_priv *priv)
 				} else {
 					new_size = rounddown_pow_of_two(ring->actual_size);
 					en_warn(priv, "Only %d buffers allocated "
-						      "reducing ring size to %d",
+						      "reducing ring size to %d\n",
 						ring->actual_size, new_size);
 					goto reduce_rings;
 				}
@@ -471,8 +471,9 @@ err_buffers:
 
 	ring_ind = priv->rx_ring_num - 1;
 err_allocator:
-	while (ring_ind >= 1) {
-		mlx4_en_destroy_allocator(priv, &priv->rx_ring[ring_ind]);
+	while (ring_ind >= 0) {
+		if (priv->rx_ring[ring_ind].use_frags)
+			mlx4_en_destroy_allocator(priv, &priv->rx_ring[ring_ind]);
 		ring_ind--;
 	}
 	return err;
-- 
1.5.3.7

