From 7b2a4a0664db41eea79b0c706fcb5d6fedb3c8eb Mon Sep 17 00:00:00 2001
From: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
Date: Wed, 17 Jun 2009 13:31:34 +0300
Subject: [PATCH 1/4] mlx4_en: Counting rx dropped when failed to create linear skb

Signed-off-by: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
---
 drivers/net/mlx4/en_rx.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/drivers/net/mlx4/en_rx.c b/drivers/net/mlx4/en_rx.c
index 9ad750b..a9698f4 100644
--- a/drivers/net/mlx4/en_rx.c
+++ b/drivers/net/mlx4/en_rx.c
@@ -855,8 +855,10 @@ int mlx4_en_process_rx_cq_skb(struct net_device *dev,
 		ring->packets++;
 
 		skb = mlx4_en_get_rx_skb(priv, rx_desc, pskb, length);
-		if (unlikely(!skb))
+		if (unlikely(!skb)){
+			priv->stats.rx_dropped++;
 			goto next;
+		}
 		skb->protocol = eth_type_trans(skb, dev);
 
 		if (likely(priv->rx_csum && cqe->checksum == 0xffff)) {
-- 
1.6.3

