mlx4_1251_get_qp_regions.patch for mlx4_fc using

Signed-off-by: Ella Podvalny <ella@mellanox.co.il>

Index: linux-2.6/drivers/net/mlx4/qp.c
===================================================================
--- linux-2.6.orig/drivers/net/mlx4/qp.c
+++ linux-2.6/drivers/net/mlx4/qp.c
@@ -331,6 +331,19 @@ void mlx4_cleanup_qp_table(struct mlx4_d
 	mlx4_bitmap_cleanup(&mlx4_priv(dev)->qp_table.bitmap);
 }
 
+int mlx4_qp_get_region(struct mlx4_dev *dev, enum mlx4_qp_region region,
+			int *base_qpn, int *cnt)
+{
+	if ((region < 0) || (region >= MLX4_NUM_QP_REGION))
+		return -EINVAL;
+
+	*base_qpn = dev->caps.reserved_qps_base[region];
+	*cnt = dev->caps.reserved_qps_cnt[region];
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(mlx4_qp_get_region);
+
 int mlx4_qp_query(struct mlx4_dev *dev, struct mlx4_qp *qp,
 		  struct mlx4_qp_context *context)
 {
Index: linux-2.6/include/linux/mlx4/qp.h
===================================================================
--- linux-2.6.orig/include/linux/mlx4/qp.h
+++ linux-2.6/include/linux/mlx4/qp.h
@@ -328,5 +328,7 @@ static inline struct mlx4_qp *__mlx4_qp_
 }
 
 void mlx4_qp_remove(struct mlx4_dev *dev, struct mlx4_qp *qp);
+int mlx4_qp_get_region(struct mlx4_dev *dev, enum mlx4_qp_region region,
+			int *base_qpn, int *cnt);
 
 #endif /* MLX4_QP_H */
