From 9c04416a9afe1289eab626c463bfc4fa8ca118f9 Mon Sep 17 00:00:00 2001
From: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
Date: Tue, 15 Dec 2009 14:12:51 +0200
Subject: [PATCH] mlx4_core: Cleanup for port configuration

Signed-off-by: Yevgeny Petrilin <yevgenyp@mellanox.co.il>
---
 drivers/net/mlx4/main.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/drivers/net/mlx4/main.c b/drivers/net/mlx4/main.c
index 75deff1..b117f55 100644
--- a/drivers/net/mlx4/main.c
+++ b/drivers/net/mlx4/main.c
@@ -1522,6 +1522,7 @@ static int __init mlx4_init(void)
 static void __exit mlx4_cleanup(void)
 {
 	mutex_lock(&drv_mutex);
+	mlx4_config_cleanup();
 	pci_unregister_driver(&mlx4_driver);
 	mutex_unlock(&drv_mutex);
 	destroy_workqueue(mlx4_wq);
-- 
1.5.3.7

