Index: ofed_kernel-fixes/drivers/infiniband/hw/mthca/mthca_qp.c
===================================================================
--- ofed_kernel-fixes.orig/drivers/infiniband/hw/mthca/mthca_qp.c	2010-03-09 12:05:07.000000000 +0200
+++ ofed_kernel-fixes/drivers/infiniband/hw/mthca/mthca_qp.c	2010-03-09 12:07:23.000000000 +0200
@@ -1490,7 +1490,9 @@ static int build_mlx_header(struct mthca
 	u16 pkey;
 
 	ib_ud_header_init(256, /* assume a MAD */
+			  1, 0, 0,
 			  mthca_ah_grh_present(to_mah(wr->wr.ud.ah)),
+			  0,
 			  &sqp->ud_header);
 
 	err = mthca_read_ah(dev, to_mah(wr->wr.ud.ah), &sqp->ud_header);
