diff --git a/drivers/infiniband/hw/nes/nes_verbs.c b/drivers/infiniband/hw/nes/nes_verbs.c
index 62f89ae..9a52ff2 100644
--- a/drivers/infiniband/hw/nes/nes_verbs.c
+++ b/drivers/infiniband/hw/nes/nes_verbs.c
@@ -3262,6 +3262,7 @@ int nes_modify_qp(struct ib_qp *ibqp, struct ib_qp_attr *attr,
 	u16 original_last_aeq;
 	u8 issue_modify_qp = 0;
 	u8 issue_disconnect = 0;
+	u8 issue_flush = 0;
 	u8 dont_wait = 0;
 
 	nes_debug(NES_DBG_MOD_QP, "QP%u: QP State=%u, cur QP State=%u,"
@@ -3377,6 +3378,12 @@ int nes_modify_qp(struct ib_qp *ibqp, struct ib_qp_attr *attr,
 				}
 				nes_debug(NES_DBG_MOD_QP, "QP%u: new state = error\n",
 						nesqp->hwqp.qp_id);
+
+				if (nesqp->iwarp_state == NES_CQP_QP_IWARP_STATE_IDLE) {
+					nesqp->flush_issued = 1;
+					issue_flush = 1;
+				}
+
 				if (nesqp->term_flags)
 					del_timer(&nesqp->terminate_timer);
 
@@ -3462,6 +3469,8 @@ int nes_modify_qp(struct ib_qp *ibqp, struct ib_qp_attr *attr,
 					" failed for QP%u.\n",
 					next_iwarp_state, nesqp->hwqp.qp_id);
 
+		if (issue_flush)
+			flush_wqes(nesdev, nesqp, NES_CQP_FLUSH_RQ | NES_CQP_FLUSH_SQ, 1);
 	}
 
 	if ((issue_modify_qp) && (nesqp->ibqp_state > IB_QPS_RTS)) {
