diff --git a/drivers/infiniband/hw/qib/qib_iba7322.c b/drivers/infiniband/hw/qib/qib_iba7322.c
index a871bad..e6e0445 100644
--- a/drivers/infiniband/hw/qib/qib_iba7322.c
+++ b/drivers/infiniband/hw/qib/qib_iba7322.c
@@ -2066,10 +2066,9 @@ static void qib_7322_handle_hwerrors(struct qib_devdata *dd, char *msg,
 	}
 	qib_stats.sps_hwerrs++;
 
-	/*
-	 * Always clear the error status register. Not doing MEMBISTFAILED hack.
-	 */
-	qib_write_kreg(dd, kr_hwerrclear, hwerrs);
+	/* Always clear the error status register, except BIST fail */
+	qib_write_kreg(dd, kr_hwerrclear, hwerrs &
+		       ~HWE_MASK(PowerOnBISTFailed));
 
 	hwerrs &= dd->cspec->hwerrmask;
 
@@ -2102,6 +2101,15 @@ static void qib_7322_handle_hwerrors(struct qib_devdata *dd, char *msg,
 		}
 	}
 
+	if (hwerrs & HWE_MASK(PowerOnBISTFailed)) {
+		isfatal = 1;
+		strlcpy(msg, "[Memory BIST test failed, "
+			"InfiniPath hardware unusable]", msgl);
+		/* ignore from now on, so disable until driver reloaded */
+		dd->cspec->hwerrmask &= ~HWE_MASK(PowerOnBISTFailed);
+		qib_write_kreg(dd, kr_hwerrmask, dd->cspec->hwerrmask);
+	}
+
 	err_decode(msg, msgl, hwerrs, qib_7322_hwerror_msgs);
 
 	/* Ignore esoteric PLL failures et al. */
@@ -2136,6 +2144,12 @@ bail:;
 static void qib_7322_init_hwerrors(struct qib_devdata *dd)
 {
 	int pidx;
+	u64 extsval;
+
+	extsval = qib_read_kreg64(dd, kr_extstatus);
+	if (!(extsval & (QIB_EXTS_MEMBIST_DISABLED |
+			 QIB_EXTS_MEMBIST_ENDTEST)))
+		qib_dev_err(dd, "MemBIST did not complete!\n");
 
 	/* never clear BIST failure, so reported on each driver load */
 	qib_write_kreg(dd, kr_hwerrclear, ~HWE_MASK(PowerOnBISTFailed));
