diff -up a/drivers/infiniband/hw/qib/qib_verbs.c b/drivers/infiniband/hw/qib/qib_verbs.c
--- a/drivers/infiniband/hw/qib/qib_verbs.c	2010-09-03 09:36:45.000000000 -0700
+++ b/drivers/infiniband/hw/qib/qib_verbs.c	2010-09-03 09:38:17.000000000 -0700
@@ -1725,6 +1725,7 @@ static int qib_query_port(struct ib_devi
 	props->sm_sl = ibp->sm_sl;
 	props->state = dd->f_iblink_state(ppd->lastibcstat);
 	props->phys_state = dd->f_ibphys_portstate(ppd->lastibcstat);
+	props->link_layer = IB_LINK_LAYER_INFINIBAND;
 	props->port_cap_flags = ibp->port_cap_flags;
 	props->gid_tbl_len = QIB_GUIDS_PER_PORT;
 	props->max_msg_sz = 0x80000000;
