diff --git a/drivers/infiniband/ulp/srp/ib_srp.c b/drivers/infiniband/ulp/srp/ib_srp.c
index e44939a..4cd8a67 100644
--- a/drivers/infiniband/ulp/srp/ib_srp.c
+++ b/drivers/infiniband/ulp/srp/ib_srp.c
@@ -526,9 +526,10 @@ static void srp_remove_req(struct srp_target_port *target, struct srp_request *r
 	list_move_tail(&req->list, &target->free_reqs);
 }
 
-static void srp_reset_req(struct srp_target_port *target, struct srp_request *req)
+static void srp_reset_req(struct srp_target_port *target,
+			  struct srp_request *req, int status)
 {
-	req->scmnd->result = DID_RESET << 16;
+	req->scmnd->result = status << 16;
 	req->scmnd->scsi_done(req->scmnd);
 	srp_remove_req(target, req);
 }
@@ -571,7 +572,7 @@ static int srp_reconnect_target(struct srp_target_port *target)
 
 	spin_lock_irq(target->scsi_host->host_lock);
 	list_for_each_entry_safe(req, tmp, &target->req_queue, list)
-		srp_reset_req(target, req);
+		srp_reset_req(target, req, DID_RESET);
 	spin_unlock_irq(target->scsi_host->host_lock);
 
 	target->rx_head	 = 0;
@@ -1431,7 +1432,7 @@ static int srp_reset_device(struct scsi_cmnd *scmnd)
 
 	list_for_each_entry_safe(req, tmp, &target->req_queue, list)
 		if (req->scmnd->device == scmnd->device)
-			srp_reset_req(target, req);
+			srp_reset_req(target, req, DID_RESET);
 
 	spin_unlock_irq(target->scsi_host->host_lock);
 
