---
 drivers/infiniband/core/ucma.c |   22 +++++++++++++++-------
 1 file changed, 15 insertions(+), 7 deletions(-)

Index: ofed_kernel/drivers/infiniband/core/ucma.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/ucma.c
+++ ofed_kernel/drivers/infiniband/core/ucma.c
@@ -1207,13 +1207,12 @@ static struct miscdevice ucma_misc = {
 	.fops	= &ucma_fops,
 };
 
-static ssize_t show_abi_version(struct device *dev,
-				struct device_attribute *attr,
-				char *buf)
+static struct class *ucma_class;
+static ssize_t show_abi_version(struct class *class_dev, char *buf)
 {
 	return sprintf(buf, "%d\n", RDMA_USER_CM_ABI_VERSION);
 }
-static DEVICE_ATTR(abi_version, S_IRUGO, show_abi_version, NULL);
+static CLASS_ATTR(abi_version, S_IRUGO, show_abi_version, NULL);
 
 static int __init ucma_init(void)
 {
@@ -1223,12 +1222,21 @@ static int __init ucma_init(void)
 	if (ret)
 		return ret;
 
-	ret = device_create_file(ucma_misc.this_device, &dev_attr_abi_version);
+	ucma_class = class_create(THIS_MODULE, "infiniband_ucma");
+	if (IS_ERR(ucma_class)) {
+		printk(KERN_ERR "rdma_ucm: couldn't create class infiniband_ucma\n");
+		goto err;
+	}
+
+	ret = class_create_file(ucma_class, &class_attr_abi_version);
 	if (ret) {
 		printk(KERN_ERR "rdma_ucm: couldn't create abi_version attr\n");
-		goto err;
+		goto err_class;
 	}
 	return 0;
+
+err_class:
+	class_destroy(ucma_class);
 err:
 	misc_deregister(&ucma_misc);
 	return ret;
@@ -1236,7 +1244,7 @@ err:
 
 static void __exit ucma_cleanup(void)
 {
-	device_remove_file(ucma_misc.this_device, &dev_attr_abi_version);
+	class_destroy(ucma_class);
 	misc_deregister(&ucma_misc);
 	idr_destroy(&ctx_idr);
 }
