Index: ofed_kernel-2.6.16_sles10_sp2/drivers/infiniband/hw/mlx4/main.c
===================================================================
--- ofed_kernel-2.6.16_sles10_sp2.orig/drivers/infiniband/hw/mlx4/main.c	2010-02-15 13:46:29.000000000 +0200
+++ ofed_kernel-2.6.16_sles10_sp2/drivers/infiniband/hw/mlx4/main.c	2010-02-15 13:51:19.000000000 +0200
@@ -1117,9 +1117,6 @@ static int mlx4_ib_netdev_event(struct n
 	struct mlx4_ib_iboe *iboe;
 	int port;
 
-	if (!net_eq(dev_net(dev), &init_net))
-		return NOTIFY_DONE;
-
 	ibdev = container_of(this, struct mlx4_ib_dev, iboe.nb);
 	iboe = &ibdev->iboe;
 
