Index: ofed_kernel-2.6.16_sles10_sp2/drivers/infiniband/ulp/sdp/sdp_main.c
===================================================================
--- ofed_kernel-2.6.16_sles10_sp2.orig/drivers/infiniband/ulp/sdp/sdp_main.c	2010-09-12 16:38:27.000000000 +0200
+++ ofed_kernel-2.6.16_sles10_sp2/drivers/infiniband/ulp/sdp/sdp_main.c	2010-09-12 16:45:25.000000000 +0200
@@ -531,7 +531,7 @@
 		return;
 	}
 
-	percpu_counter_dec(sk->sk_prot->orphan_count);
+	atomic_dec(sk->sk_prot->orphan_count);
 	ssk->destructed_already = 1;
 
 	down_read(&device_removal_lock);
@@ -2646,9 +2646,9 @@
 		sk->sk_data_ready(sk, 0);
 }
 
-static struct percpu_counter *sockets_allocated;
+static atomic_t sockets_allocated;
 static atomic_t memory_allocated;
-static struct percpu_counter *orphan_count;
+static atomic_t orphan_count;
 static int memory_pressure;
 struct proto sdp_proto = {
         .close       = sdp_close,
@@ -2666,8 +2666,10 @@
         .get_port    = sdp_get_port,
 	/* Wish we had this: .listen   = sdp_listen */
 	.enter_memory_pressure = sdp_enter_memory_pressure,
+	.sockets_allocated = &sockets_allocated,
 	.memory_allocated = &memory_allocated,
 	.memory_pressure = &memory_pressure,
+	.orphan_count = &orphan_count,
         .sysctl_mem             = sysctl_tcp_mem,
         .sysctl_wmem            = sysctl_tcp_wmem,
         .sysctl_rmem            = sysctl_tcp_rmem,
@@ -2897,20 +2899,6 @@
 	spin_lock_init(&sock_list_lock);
 	spin_lock_init(&sdp_large_sockets_lock);
 
-	sockets_allocated = kzalloc(sizeof(*sockets_allocated), GFP_KERNEL);
-	if (!sockets_allocated)
-		goto no_mem_sockets_allocated;
-
-	orphan_count = kzalloc(sizeof(*orphan_count), GFP_KERNEL);
-	if (!orphan_count)
-		goto no_mem_orphan_count;
-
-	percpu_counter_init(sockets_allocated, 0);
-	percpu_counter_init(orphan_count, 0);
-
-	sdp_proto.sockets_allocated = sockets_allocated;
-	sdp_proto.orphan_count = orphan_count;
-
 	rx_comp_wq = create_workqueue("rx_comp_wq");
 	if (!rx_comp_wq)
 		goto no_mem_rx_wq;
@@ -2946,10 +2934,6 @@
 no_mem_sdp_wq:
 	destroy_workqueue(rx_comp_wq);
 no_mem_rx_wq:
-	kfree(orphan_count);
-no_mem_orphan_count:
-	kfree(sockets_allocated);
-no_mem_sockets_allocated:
 	return rc;
 }
 
@@ -2958,9 +2942,9 @@
 	sock_unregister(PF_INET_SDP);
 	proto_unregister(&sdp_proto);
 
-	if (percpu_counter_read_positive(orphan_count))
-		printk(KERN_WARNING "%s: orphan_count %lld\n", __func__,
-		       percpu_counter_read_positive(orphan_count));
+	if (atomic_read(&orphan_count))
+		printk(KERN_WARNING "%s: orphan_count %d\n", __func__,
+		       atomic_read(&orphan_count));
 
 	destroy_workqueue(rx_comp_wq);
 	destroy_workqueue(sdp_wq);
@@ -2974,12 +2958,6 @@
 	sdp_proc_unregister();
 
 	ib_unregister_client(&sdp_client);
-
-	percpu_counter_destroy(sockets_allocated);
-	percpu_counter_destroy(orphan_count);
-
-	kfree(orphan_count);
-	kfree(sockets_allocated);
 }
 
 module_init(sdp_init);
Index: ofed_kernel-2.6.16_sles10_sp2/drivers/infiniband/ulp/sdp/sdp.h
===================================================================
--- ofed_kernel-2.6.16_sles10_sp2.orig/drivers/infiniband/ulp/sdp/sdp.h	2010-09-12 16:38:27.000000000 +0200
+++ ofed_kernel-2.6.16_sles10_sp2/drivers/infiniband/ulp/sdp/sdp.h	2010-09-12 16:44:18.000000000 +0200
@@ -112,7 +112,7 @@
 		sdp_dbg(sk, "%s:%d - sock_put(" SOCK_REF_ALIVE \
 			") - refcount = %d from withing sk_common_release\n",\
 			__func__, __LINE__, atomic_read(&(sk)->sk_refcnt));\
-		percpu_counter_inc((sk)->sk_prot->orphan_count);\
+		atomic_inc((sk)->sk_prot->orphan_count);\
 		sk_common_release(sk); \
 } while (0)
 
