---
 drivers/infiniband/core/addr.c |    6 ++++++
 1 file changed, 6 insertions(+)

Index: ofed_kernel/drivers/infiniband/core/addr.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/addr.c
+++ ofed_kernel/drivers/infiniband/core/addr.c
@@ -32,6 +32,12 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
+#include <linux/if.h>
+#include <linux/netdevice.h>
+#include <linux/timer.h>
+
+/* XXX this define is the same for all architectures in RHAS4.0. */
+#define ECANCELED 125
 
 #include <linux/mutex.h>
 #include <linux/inetdevice.h>
