---
 drivers/infiniband/ulp/ipoib/ipoib_multicast.c |    4 ++++
 1 file changed, 4 insertions(+)

Index: ofed_kernel/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
+++ ofed_kernel/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
@@ -796,6 +796,10 @@ void ipoib_mcast_restart_task(struct wor
 
 		memcpy(mgid.raw, mclist->dmi_addr + 4, sizeof mgid);
 
+		/* Add in the P_Key */
+		mgid.raw[4] = (priv->pkey >> 8) & 0xff;
+		mgid.raw[5] = priv->pkey & 0xff;
+
 		mcast = __ipoib_mcast_find(dev, &mgid);
 		if (!mcast || test_bit(IPOIB_MCAST_FLAG_SENDONLY, &mcast->flags)) {
 			struct ipoib_mcast *nmcast;
