---
 drivers/infiniband/hw/cxgb3/iwch_provider.c |    7 -------
 1 file changed, 7 deletions(-)

Index: ofed_kernel/drivers/infiniband/hw/cxgb3/iwch_provider.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/cxgb3/iwch_provider.c
+++ ofed_kernel/drivers/infiniband/hw/cxgb3/iwch_provider.c
@@ -358,13 +358,6 @@ static int iwch_mmap(struct ib_ucontext 
 	    (addr < (rdev_p->rnic_info.udbell_physbase +
 		       rdev_p->rnic_info.udbell_len))) {
 
-		/*
-		 * Map T3 DB register.
-		 */
-		if (vma->vm_flags & VM_READ) {
-			return -EPERM;
-		}
-
 		vma->vm_page_prot = pgprot_noncached(vma->vm_page_prot);
 		vma->vm_flags |= VM_DONTCOPY | VM_DONTEXPAND;
 		vma->vm_flags &= ~VM_MAYREAD;
