---
 drivers/infiniband/hw/mlx4/wc.c |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

Index: ofed_kernel/drivers/infiniband/hw/mlx4/wc.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/hw/mlx4/wc.c
+++ ofed_kernel/drivers/infiniband/hw/mlx4/wc.c
@@ -111,7 +111,7 @@ static int read_and_modify_pat(void)
 	preempt_disable();
 	rd_old_pat(&ret);
 	if (!ret)
-		smp_call_function(rd_old_pat, &ret, 1);
+		smp_call_function(rd_old_pat, &ret, 1, 1);
 	if (ret)
 		goto out;
 
@@ -119,7 +119,7 @@ static int read_and_modify_pat(void)
 	if (ret)
 		goto out;
 
-	smp_call_function(wr_new_pat, &ret, 1);
+	smp_call_function(wr_new_pat, &ret, 1, 1);
 	BUG_ON(ret); /* have inconsistent PAT state */
 out:
 	preempt_enable();
@@ -133,7 +133,7 @@ static int restore_pat(void)
 	preempt_disable();
 	wr_old_pat(&ret);
 	if (!ret) {
-		smp_call_function(wr_old_pat, &ret, 1);
+		smp_call_function(wr_old_pat, &ret, 1, 1);
 		BUG_ON(ret); /* have inconsistent PAT state */
 	}
 
