---
 drivers/net/mlx4/eq.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: ofed_kernel/drivers/net/mlx4/eq.c
===================================================================
--- ofed_kernel.orig/drivers/net/mlx4/eq.c
+++ ofed_kernel/drivers/net/mlx4/eq.c
@@ -532,7 +532,7 @@ int mlx4_map_eq_icm(struct mlx4_dev *dev
 		return -ENOMEM;
 	priv->eq_table.icm_dma  = pci_map_page(dev->pdev, priv->eq_table.icm_page, 0,
 					       PAGE_SIZE, PCI_DMA_BIDIRECTIONAL);
-	if (pci_dma_mapping_error(dev->pdev, priv->eq_table.icm_dma)) {
+	if (pci_dma_mapping_error(priv->eq_table.icm_dma)) {
 		__free_page(priv->eq_table.icm_page);
 		return -ENOMEM;
 	}
