diff --git a/drivers/net/mlx4/main.c b/drivers/net/mlx4/main.c
index 1edce39..86e56dd 100644
--- a/drivers/net/mlx4/main.c
+++ b/drivers/net/mlx4/main.c
@@ -530,7 +530,7 @@ out:
 	return err ? err : count;
 }
 
-static ssize_t trigger_port(struct device *dev, struct device_attribute *attr,
+static ssize_t trigger_port(struct device *dev,
 			    const char *buf, size_t count)
 {
 	struct pci_dev *pdev = to_pci_dev(dev);
