Index: ofed_kernel-2.6.9_U8/drivers/infiniband/ulp/ipoib/ipoib_main.c
===================================================================
--- ofed_kernel-2.6.9_U8.orig/drivers/infiniband/ulp/ipoib/ipoib_main.c	2010-08-30 15:00:19.000000000 +0300
+++ ofed_kernel-2.6.9_U8/drivers/infiniband/ulp/ipoib/ipoib_main.c	2010-08-30 16:03:54.721530666 +0300
@@ -1294,7 +1294,6 @@ static struct net_device *ipoib_add_port
 		goto alloc_mem_failed;
 
 	SET_NETDEV_DEV(priv->dev, hca->dma_device);
-	priv->dev->dev_id = port - 1;
 
 	if (!ib_query_port(hca, port, &attr))
 		priv->max_ib_mtu = ib_mtu_enum_to_int(attr.max_mtu);
