---
 drivers/infiniband/core/cma.c |    5 +----
 1 file changed, 1 insertion(+), 4 deletions(-)

Index: ofed_kernel/drivers/infiniband/core/cma.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/cma.c
+++ ofed_kernel/drivers/infiniband/core/cma.c
@@ -2835,13 +2835,10 @@ static int cma_netdev_callback(struct no
 	struct rdma_id_private *id_priv;
 	int ret = NOTIFY_DONE;
 
-	if (dev_net(ndev) != &init_net)
-		return NOTIFY_DONE;
-
 	if (event != NETDEV_BONDING_FAILOVER)
 		return NOTIFY_DONE;
 
-	if (!(ndev->flags & IFF_MASTER) || !(ndev->priv_flags & IFF_BONDING))
+	if (!(ndev->flags & IFF_MASTER))
 		return NOTIFY_DONE;
 
 	mutex_lock(&lock);
