---
 drivers/infiniband/ulp/ipoib/ipoib_multicast.c |    4 ++++
 1 file changed, 4 insertions(+)

Index: ofed_kernel-2.6.24/drivers/infiniband/ulp/ipoib/ipoib_multicast.c
===================================================================
--- ofed_kernel-2.6.24.orig/drivers/infiniband/ulp/ipoib/ipoib_multicast.c	2010-03-11 10:12:12.000000000 +0200
+++ ofed_kernel-2.6.24/drivers/infiniband/ulp/ipoib/ipoib_multicast.c	2010-03-11 10:27:00.000000000 +0200
@@ -778,10 +778,10 @@ static int ipoib_mcast_addr_is_valid(con
 	if (addrlen != INFINIBAND_ALEN)
 		return 0;
 	/* reserved QPN, prefix, scope */
-	if (memcmp(addr, broadcast, 6))
+	if (memcmp(addr, broadcast, 5))
 		return 0;
-	/* signature lower, pkey */
-	if (memcmp(addr + 7, broadcast + 7, 3))
+	/* signature lower */
+	if (addr[7] != broadcast[7])
 		return 0;
 	return 1;
 }
@@ -826,6 +826,11 @@ void ipoib_mcast_restart_task(struct wor
 
 		memcpy(mgid.raw, mclist->dmi_addr + 4, sizeof mgid);
 
+		/* update scope */
+		mgid.raw[1] = 0x10 | (dev->broadcast[5] & 0xF);
+		/* Add in the P_Key */
+		mgid.raw[4] = dev->broadcast[8];
+		mgid.raw[5] = dev->broadcast[9];
 		mcast = __ipoib_mcast_find(dev, &mgid);
 		if (!mcast || test_bit(IPOIB_MCAST_FLAG_SENDONLY, &mcast->flags)) {
 			struct ipoib_mcast *nmcast;
