Index: ofed_kernel/drivers/infiniband/core/ucm.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/ucm.c	2009-12-16 16:37:11.000000000 +0200
+++ ofed_kernel/drivers/infiniband/core/ucm.c	2009-12-17 10:48:44.000000000 +0200
@@ -43,6 +43,7 @@
 #include <linux/cdev.h>
 #include <linux/idr.h>
 #include <linux/mutex.h>
+#include <linux/sched.h>
 
 #include <asm/uaccess.h>
 
Index: ofed_kernel/drivers/infiniband/core/uverbs_main.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/uverbs_main.c	2009-12-17 10:05:07.000000000 +0200
+++ ofed_kernel/drivers/infiniband/core/uverbs_main.c	2009-12-17 10:50:09.000000000 +0200
@@ -43,6 +43,7 @@
 #include <linux/file.h>
 #include <linux/mount.h>
 #include <linux/cdev.h>
+#include <linux/sched.h>
 
 #include <asm/uaccess.h>
 
Index: ofed_kernel/drivers/infiniband/core/ucma.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/ucma.c	2009-12-16 16:37:11.000000000 +0200
+++ ofed_kernel/drivers/infiniband/core/ucma.c	2009-12-17 10:53:59.000000000 +0200
@@ -38,6 +38,7 @@
 #include <linux/in.h>
 #include <linux/in6.h>
 #include <linux/miscdevice.h>
+#include <linux/sched.h>
 
 #include <rdma/rdma_user_cm.h>
 #include <rdma/ib_marshall.h>

Index: ofed_kernel/drivers/infiniband/core/iwcm.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/iwcm.c	2009-12-17 12:27:11.000000000 +0200
+++ ofed_kernel/drivers/infiniband/core/iwcm.c	2009-12-17 12:32:25.000000000 +0200
@@ -43,6 +43,7 @@
 #include <linux/spinlock.h>
 #include <linux/workqueue.h>
 #include <linux/completion.h>
+#include <linux/sched.h>
 
 #include <rdma/iw_cm.h>
 #include <rdma/ib_addr.h>
Index: ofed_kernel/drivers/infiniband/core/user_mad.c
===================================================================
--- ofed_kernel.orig/drivers/infiniband/core/user_mad.c	2010-03-22 13:15:06.000000000 +0200
+++ ofed_kernel/drivers/infiniband/core/user_mad.c	2010-03-22 13:16:08.000000000 +0200
@@ -45,6 +45,7 @@
 #include <linux/kref.h>
 #include <linux/compat.h>
 #include <linux/semaphore.h>
+#include <linux/sched.h>
 
 #include <asm/uaccess.h>
 
