---
 net/rds/connection.c |    3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

Index: ofed_kernel/net/rds/connection.c
===================================================================
--- ofed_kernel.orig/net/rds/connection.c
+++ ofed_kernel/net/rds/connection.c
@@ -51,8 +51,7 @@ static struct kmem_cache *rds_conn_slab;
 static struct hlist_head *rds_conn_bucket(__be32 laddr, __be32 faddr)
 {
 	/* Pass NULL, don't need struct net for hash */
-	unsigned long hash = inet_ehashfn(NULL,
-					  be32_to_cpu(laddr), 0, 
+	unsigned long hash = inet_ehashfn(be32_to_cpu(laddr), 0,
 					  be32_to_cpu(faddr), 0);
 	return &rds_conn_hash[hash & RDS_CONNECTION_HASH_MASK];
 }
