Index: ofa_kernel-1.5.1/drivers/infiniband/debug/memtrack.c
===================================================================
--- ofa_kernel-1.5.1.orig/drivers/infiniband/debug/memtrack.c	2010-07-29 10:45:51.000000000 -0700
+++ ofa_kernel-1.5.1/drivers/infiniband/debug/memtrack.c	2010-08-01 12:38:21.000000000 -0700
@@ -440,7 +440,6 @@
                         }
                         else {
                                 proc_ent->proc_fops = &memtrack_proc_fops;
-                                proc_ent->owner = THIS_MODULE;
                         }
                 }
         }
