---
 drivers/scsi/iscsi_tcp.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: ofed_kernel/drivers/scsi/iscsi_tcp.c
===================================================================
--- ofed_kernel.orig/drivers/scsi/iscsi_tcp.c
+++ ofed_kernel/drivers/scsi/iscsi_tcp.c
@@ -2027,7 +2027,7 @@ iscsi_tcp_conn_get_param(struct iscsi_cl
 {
 	struct iscsi_conn *conn = cls_conn->dd_data;
 	struct iscsi_tcp_conn *tcp_conn = conn->dd_data;
-	struct inet_sock *inet;
+	struct inet_opt *inet;
 	struct ipv6_pinfo *np;
 	struct sock *sk;
 	int len;
