--- a/drivers/infiniband/hw/ipath/ipath_wc_pat.c.orig	2008-11-04 13:46:56.557658000 -0800
+++ b/drivers/infiniband/hw/ipath/ipath_wc_pat.c	2008-11-04 13:47:50.521639000 -0800
@@ -182,7 +182,7 @@
 	preempt_disable();
 	rd_old_pat(&ret);
 	if (!ret)
-		smp_call_function(rd_old_pat, &ret, 1);
+		smp_call_function(rd_old_pat, &ret, 1, 1);
 	if (ret)
 		goto out;
 
@@ -193,7 +193,7 @@
 	if (ret)
 		goto out;
 
-	smp_call_function(wr_new_pat, &ret, 1);
+	smp_call_function(wr_new_pat, &ret, 1, 1);
 	BUG_ON(ret); /* have inconsistent PAT state */
 out:
 	preempt_enable();
@@ -207,7 +207,7 @@
 	preempt_disable();
 	wr_old_pat(&ret);
 	if (!ret) {
-		smp_call_function(wr_old_pat, &ret, 1);
+		smp_call_function(wr_old_pat, &ret, 1, 1);
 		BUG_ON(ret); /* have inconsistent PAT state */
 	}
 
